### DISCUSSION TITLE
Collaborate on creating raspberry pi version that can additionally run emulators

### DISCUSSION THREAD

**Discussion by reubenajohnston** [sanni/cartreader#374] [Ideas & Suggestions]
Sanni,

I am considering making an open source device like the retrode2 (dumps roms and emulates them) and thought it would be a good idea to base the ROM dumping portions off this sanni/cartreader project.  If you are interested in collaborating, please reach out to me.  My thought is to base it off a regular raspberry pi 4.  

Regards,
Reub (reubenajohnston)
reuben.a.johnston@gmail.com

**Comment by Dakkaron** [comment]
Not to be a downer, but that sounds quite hard to port. The Pi 4 doesn't have enough pins. It's limited to 3.3V on the GPIO pins and timing is going to be rather hard, since the Pi has an OS in between which means, you can't even get a reliable kilohertz range clock out of a GPIO pin...

I guess, you'll end up with an ATMega (or similar microcontroller) between the Pi and the Cartridge. At this point you can just strap the cartreader to your Pi and stream the dumped cartridge to the Pi.

Then again, why do that? Dumping can take quite some time, so running the games from a dump (e.g. dumping to the Pi and running the emulator from the dump) is going to be rather boring. So you could just dump the game once in advance and just run the emulator directly from them dump. That way, also the cartridge (and reader) is not needed for running the game. So might as well just dump the game on the cartreader, transfer the dump to the Pi and run it there from the file.

The other approach would be to not dump the game, but instead "connect" the emulator directly to the cartridge. So that the emulator uses the cartridge the same way a real console would use the cartridge. This is going to be really difficult to pull off. First, you will need to modify an emulator to work that way. Second, you need to reduce latency to next to nothing and your timings are going to be really critical. The Pi won't be able to do the cartridge interface on it's own, so you need an Arduino in between. But USB access to the cartridge will have far too much latency. Serial over USB is out for the same reason. UART or SPI could theoretically work, but that's unchartered territory afaik...

At any rate, this is not going to be easy at all, while there exists a very mature alternative (dump and play the ROM on an emulator) that is so mature that we've been doing that for more than 25 years. And that solution doesn't even require you to make anything different than what we already have... It's literally just using the cartreader and an emulator the way they were meant to.

**Reply by reubenajohnston** [reply]
It wasn't my intention to start an architectural or feature discussion.  The point was to see whether there was interest in collaborating.  There are many features that I have in mind that will be better suited with RPI CM4 (or equivalent) + FPGA.  Porting cartreader features will be trivial (my opinion).  Some of the other features that I have in mind, are definitely not trivial but could potentially leverage other open source (e.g., mister).  
Reub

**Reply by Dakkaron** [reply]
I am not Sanni and I obviously don't speak for him. I wasn't talking about features or architecture either. It was more of a question regarding the whole concept.

Basically, what I am asking is:

- Is the concept feasible?
- What benefit does the concept have over the traditional way of handling things?
- How does your concept benefit from the cartreader specifically?
- What parts of the cartreader make sense to be ported to your concept in general?

(Bonus question: What's the point of using a Pi and a FPGA in the same emulation device? They usually fill the same role in an emulation device.)
